plugins/experimental/alt_svc/alt_svc.cc

```diff
@@ -0,0 +1,81 @@
+/**
+  @file
+  @brief TODO documentation
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include <atscppapi/GlobalPlugin.h>
+#include <atscppapi/PluginInit.h>
+#include "ip_host_map.h"
+#include "default.h"
+
+using namespace atscppapi;
+using namespace std;
+
+// plugin registration info
+static char vendor_name[]   = "Yahoo! Inc.";
+static char support_email[] = "ats-devel@yahoo-inc.com";
+
+namespace
+{
+GlobalPlugin *plugin;
+}
+
+class AltSvcHeaderPlugin : public GlobalPlugin
+{
+private:
+  IpHostMap *_hostmap;
+
+public:
+  AltSvcHeaderPlugin(IpHostMap *hostmap)
+  {
+    // Only if the map is initialized properly do we then add the hook.
+    registerHook(HOOK_SEND_RESPONSE_HEADERS);
+    this->_hostmap = hostmap;
+  }
+
+  void
+  handleSendResponseHeaders(Transaction &transaction)
+  {
+    char buff[70];
+    const sockaddr *client_address = transaction.getClientAddress();
+    TS_DEBUG(PLUGIN_NAME, "Looking for hostname for IP %s", ats_ip_ntop(client_address, buff, 70));
+    string host = this->_hostmap->findHostForIP(client_address, "<HOST GOES HERE LATER>");
+    if (!host.empty()) {
+      TS_DEBUG(PLUGIN_NAME, "Found hostname %s", host.c_str());
+      transaction.getClientResponse().getHeaders().append("Alt-Svc", "h2=\"" + host + ":443\"");
```

Comment:
Are the protocol name and parameters going to be configurable? It would be great if we could support other protocols and parameters in the future without changing the configuration file format.

Two parameters are defined on RFC7838.
https://www.iana.org/assignments/http-alt-svc-parameters/http-alt-svc-parameters.xhtml

Some new parameters going to be introduced in near future for QUIC, they are still draft though.
https://tools.ietf.org/html/draft-ietf-quic-http-04#section-2.1